File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int answer = 0;
        int start = 0;
        char [] chars = s.toCharArray();
        Map<Character, Integer> map = new HashMap<>();
        for(int end = 0; end < chars.length; end++){
            if(map.containsKey(chars[end])){
                start = Math.max(map.get(chars[end]) + 1, start);
            }
            answer = Math.max(answer, end - start + 1);
            map.put(chars[end], end);
        }
        return answer;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var answer = 0
        var start = 0
        val map = HashMap<Char, Int>()
        val c = s.toCharArray()
        for(end in 0 until c.size){
            if(map.containsKey(c[end])){
                start = maxOf(start, map[c[end]]!! + 1)
            }
            answer = maxOf(answer, end - start + 1);
            map[c[end]] = end;
        }
        return answer;
    }
}